pkg/distributor/distributor.go

```diff
@@ -876,74 +900,87 @@ func detectLogLevelFromLogEntry(entry logproto.Entry, structuredMetadata labels.
 		} else if otlpSeverityNumber <= int(plog.SeverityNumberFatal4) {
 			return logLevelFatal
 		}
-		return logLevelInfo
+		return logLevelUnknown
 	}
 
 	return extractLogLevelFromLogLine(entry.Line)
 }
 
 func extractLogLevelFromLogLine(log string) string {
-	// check for log levels in known log formats to avoid any false detection
+	var v string
+	if isJSON(log) {
+		v = getValueUsingJSONParser(log)
+	} else {
+		v = getValueUsingLogfmtParser(log)
+	}
+
+	switch strings.ToLower(v) {
+	case "trace", "trc":
+		return logLevelTrace
+	case "debug", "dbg":
+		return logLevelDebug
+	case "info", "inf":
+		return logLevelInfo
+	case "warn", "wrn":
+		return logLevelWarn
+	case "error", "err":
+		return logLevelError
+	case "critical":
+		return logLevelCritical
+	case "fatal":
+		return logLevelFatal
+	default:
+		return detectLevelFromLogLine(log)
+	}
+}
+
+func getValueUsingLogfmtParser(line string) string {
```

Comment:
can you return a slice of bytes here to avoid alloc ? I think library both can ?

Reply:
tried returning a `[]byte` but that always is causing more bytes allocated than returning `string`. 

**returning []byte**

![Screenshot 2024-05-06 at 12 07 00](https://github.com/grafana/loki/assets/660283/b2a47d79-ea87-473b-897b-91bd003d66dc)

**returning string**

![Screenshot 2024-05-06 at 12 07 49](https://github.com/grafana/loki/assets/660283/009f3597-7715-4325-ae84-1369d851730a)


My best guess is because string is immutable, a fixed capacity is allocated on the heap as opposed to []byte